src/rpc/methods/f3/types.rs

```diff
@@ -135,3 +135,102 @@ impl PartialOrd for F3PowerEntry {
         Some(self.cmp(other))
     }
 }
+
+#[derive(Debug, Default)]
+pub struct F3LeaseManager(RwLock<HashMap<u64, chrono::DateTime<chrono::Utc>>>);
+
+impl F3LeaseManager {
+    pub fn get_active_participants(&self) -> HashSet<u64> {
+        let now = chrono::Utc::now();
+        self.0
+            .read()
+            .iter()
+            .filter_map(|(id, expire)| if expire > &now { Some(*id) } else { None })
+            .collect()
+    }
+
+    pub fn upsert_defensive(
+        &self,
+        id: u64,
+        new_lease_expiration: chrono::DateTime<chrono::Utc>,
+        old_lease_expiration: chrono::DateTime<chrono::Utc>,
+    ) -> anyhow::Result<bool> {
+        if new_lease_expiration > chrono::Utc::now() + chrono::Duration::minutes(5) {
+            anyhow::bail!("F3 participation lease cannot be over 5 mins");
+        } else if new_lease_expiration < chrono::Utc::now() {
+            anyhow::bail!("F3 participation lease is in the past");
+        }
+
+        // if the old lease is expired just insert a new one
+        if old_lease_expiration < chrono::Utc::now() {
+            self.0.write().insert(id, new_lease_expiration);
+            return Ok(true);
+        }
+
+        let Some(old_lease_expiration_in_record) = self.0.read().get(&id).cloned() else {
+            // we don't know about it, don't start a new lease
+            return Ok(false);
+        };
+        if old_lease_expiration_in_record != old_lease_expiration {
+            // the lease we know about does not match and because the old lease is not expired
+            // we should not allow for new lease
+            return Ok(false);
+        }
```

Comment:
This seems to be the cheapest operation - should it be moved earlier?

Reply:
I think so, otherwise it seems buggy that any old_lease in the past can bypass all the checks. I gonna ask around if this is a bug in lotus

Reply:
I will fix this in a follow-up PR if the issue is confirmed. For now the logic matches lotus.